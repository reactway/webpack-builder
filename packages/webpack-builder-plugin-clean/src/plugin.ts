import CleanWebpackPlugin from "clean-webpack-plugin";
import { Plugin } from "@reactway/webpack-builder";
import { Options } from "./plugin-options";

interface CleanPluginOptions {
    options?: Partial<Options>;
}

export const CleanPlugin: Plugin<CleanPluginOptions> = (config, projectDirectory) => webpack => {
    if (webpack.plugins == null) {
        webpack.plugins = [];
    }

    let cleanPluginConfig: Partial<Options> = {};
    if (config != null && config.options != null) {
        cleanPluginConfig = config.options;
    }

    webpack.plugins.push(new CleanWebpackPlugin({ ...cleanPluginConfig }));
    return webpack;
};
