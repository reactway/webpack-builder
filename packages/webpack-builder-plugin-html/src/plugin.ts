import HtmlWebpackPlugin from "html-webpack-plugin";
import HtmlWebpackTemplate from "html-webpack-template";
import { Plugin } from "@reactway/webpack-builder";

interface HtmlPluginOptions {
    options?: HtmlWebpackPlugin.Options;
}

export const HtmlPlugin: Plugin<HtmlPluginOptions> = (config, projectDirectory) => webpack => {
    if (webpack.plugins == null) {
        webpack.plugins = [];
    }

    let htmlPluginOptions: HtmlWebpackPlugin.Options | undefined = {
        inject: false,
        template: HtmlWebpackTemplate,
        baseHref: "/",
        appMountIds: ["root"],
        meta: [
            {
                name: "viewport",
                content: "width=device-width, initial-scale=1"
            }
        ]
    };

    if (config != null) {
        htmlPluginOptions = config.options;
    }

    // ts-jest: Throws error when it cannot resolve project root, because we are using Rush/pnpm.
    // tslint:disable-next-line:no-any
    webpack.plugins.push(new HtmlWebpackPlugin(htmlPluginOptions) as any);
    return webpack;
};
