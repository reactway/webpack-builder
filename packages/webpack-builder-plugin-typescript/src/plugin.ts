import ForkTsCheckerWebpackPlugin from "fork-ts-checker-webpack-plugin";
import { TsconfigPathsPlugin } from "tsconfig-paths-webpack-plugin/lib";
import { Options as TsconfigPathsPluginOptions } from "tsconfig-paths-webpack-plugin/lib/options";
import { loadTsconfig, Tsconfig } from "tsconfig-paths/lib/tsconfig-loader";
import upath from "upath";
import * as fs from "fs-extra";
import { Plugin } from "@reactway/webpack-builder";
import { ForkTsCheckerWebpackPluginOptions } from "./plugin-options";

import "ts-loader";

// Extensions.
const TS_EXTENSION: string = ".ts";
const TSX_EXTENSION: string = ".tsx";
const JS_EXTENSION: string = ".js";
const JSX_EXTENSION: string = ".jsx";

// Tsconfig.
export const TS_CONFIG_NAME: string = "tsconfig.json";
const DEFAULT_TS_CONFIG_LOCATION: string = upath.resolve(__dirname, `../assets/${TS_CONFIG_NAME}`);

// TsLint.
export const TSLINT_CONFIG_NAME: string = "tslint.json";
const DEFAULT_TSLINT_CONFIG_LOCATION: string = upath.resolve(__dirname, `../assets/${TSLINT_CONFIG_NAME}`);

interface TypeScriptPluginOptions {
    forkTsCheckerOptions?: Partial<ForkTsCheckerWebpackPluginOptions>;
    tsconfigPathsPluginOptions?: Partial<TsconfigPathsPluginOptions>;
}

export const TypeScriptPlugin: Plugin<TypeScriptPluginOptions> = (config, projectDirectory) => {
    const fullTsconfigLocation = upath.resolve(projectDirectory, TS_CONFIG_NAME);
    let baseURLExist: boolean = false;

    try {
        checkTsConfig(projectDirectory);
    } catch (error) {
        console.error(`Failed while initiating "${TS_CONFIG_NAME}".`, error);
    }

    try {
        checkTslintConfig(projectDirectory);
    } catch (error) {
        console.error(`Failed while initiating "${TSLINT_CONFIG_NAME}".`, error);
    }

    const tsConfig: Tsconfig | undefined = loadTsconfig(fullTsconfigLocation);

    return webpack => {
        if (webpack.plugins == null) {
            webpack.plugins = [];
        }

        webpack.devtool = "inline-source-map";

        let forkTsConfig: Partial<ForkTsCheckerWebpackPluginOptions> = {};
        if (config != null && config.forkTsCheckerOptions != null) {
            forkTsConfig = config.forkTsCheckerOptions;
        }

        webpack.plugins.push(
            new ForkTsCheckerWebpackPlugin({
                checkSyntacticErrors: true,
                tslint: true,
                ...forkTsConfig
            })
        );

        if (webpack.module == null) {
            webpack.module = {
                rules: []
            };
        }

        webpack.module.rules.push({
            test: /\.tsx?$/,
            use: [
                {
                    loader: "babel-loader",
                    options: {
                        babelrc: true,
                        plugins: ["syntax-dynamic-import"]
                    }
                },
                {
                    loader: "ts-loader",
                    options: {
                        // disable type checker - we will use it in fork plugin
                        // IMPORTANT! use happyPackMode mode to speed-up compilation and reduce errors reported to webpack
                        happyPackMode: true,
                        transpileOnly: true
                    }
                }
            ],
            exclude: /node_modules/
        });

        webpack.module.rules.push({
            test: /\.jsx?$/,
            use: [
                {
                    loader: "babel-loader",
                    options: {
                        presets: ["@babel/preset-env"]
                    }
                }
            ],
            include: /node_modules/
        });

        if (webpack.resolve == null) {
            webpack.resolve = {};
        }

        if (webpack.resolve.plugins == null) {
            webpack.resolve.plugins = [];
        }

        if (
            tsConfig != null &&
            tsConfig.compilerOptions != null &&
            tsConfig.compilerOptions.baseUrl != null &&
            tsConfig.compilerOptions.baseUrl.trim().length !== 0
        ) {
            let configToSpread = {};
            if (config != null && config.tsconfigPathsPluginOptions != null) {
                configToSpread = config.tsconfigPathsPluginOptions;
            }

            const defaultOptions: Partial<TsconfigPathsPluginOptions> = {
                configFile: fullTsconfigLocation,
                ...configToSpread
            };

            webpack.resolve.plugins.push(new TsconfigPathsPlugin(defaultOptions));

            baseURLExist = true;
        }

        if (!baseURLExist && config != null && config.tsconfigPathsPluginOptions != null) {
            throw new Error(`Cannot add tsconfigPathsPluginOptions because baseUrl do not exist at ${TS_CONFIG_NAME}`);
        }

        if (webpack.resolve.extensions == null) {
            webpack.resolve.extensions = [];
        }

        if (webpack.resolve.extensions.indexOf(TS_EXTENSION) === -1) {
            webpack.resolve.extensions.push(TS_EXTENSION);
        }

        if (tsConfig != null && tsConfig.compilerOptions != null) {
            // HACK: To get other compiler options properties.
            // tslint:disable-next-line:no-any
            const tsConfigCompilerOptions = tsConfig.compilerOptions as any;
            if (tsConfigCompilerOptions.jsx != null) {
                if (webpack.resolve.extensions.indexOf(TSX_EXTENSION) === -1) {
                    webpack.resolve.extensions.push(TSX_EXTENSION);
                }
                if (webpack.resolve.extensions.indexOf(JSX_EXTENSION) === -1) {
                    webpack.resolve.extensions.push(JSX_EXTENSION);
                }
            }
        }

        if (webpack.resolve.extensions.indexOf(JS_EXTENSION) === -1) {
            webpack.resolve.extensions.push(JS_EXTENSION);
        }

        return webpack;
    };
};

export function checkTsConfig(projectDirectory: string): void {
    const configLocation = upath.resolve(projectDirectory, TS_CONFIG_NAME);

    if (!fs.pathExistsSync(configLocation)) {
        console.info(`File "${TS_CONFIG_NAME}" not found at ${configLocation}. Creating...`);
        fs.copySync(DEFAULT_TS_CONFIG_LOCATION, configLocation);
        console.info("Created.");
    }
}

export function checkTslintConfig(projectDirectory: string): void {
    const configLocation = upath.resolve(projectDirectory, TSLINT_CONFIG_NAME);

    if (!fs.pathExistsSync(configLocation)) {
        console.info(`File "${TSLINT_CONFIG_NAME}" not found at ${configLocation}. Creating...`);
        fs.copySync(DEFAULT_TSLINT_CONFIG_LOCATION, configLocation);
        console.info("Created.");
    }
}
