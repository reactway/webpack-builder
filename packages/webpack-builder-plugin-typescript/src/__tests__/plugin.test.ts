jest.mock("fork-ts-checker-webpack-plugin");
jest.mock("tsconfig-paths-webpack-plugin");
__dirname = "./src/__tests__/";

jest.mock("upath", () => ({
    ...jest.requireActual("upath"),
    resolve: (...pathSegments: string[]) => pathSegments.join("//"),
    join: (...pathSegments: string[]) => pathSegments.join("//")
}));

import { Builder, Configuration } from "@reactway/webpack-builder";
import upath from "upath";
import * as fs from "fs-extra";
// Plugin
import { TypeScriptPlugin } from "../plugin";

let SAMPLE_CONFIGURATION: Configuration = {};
const TEST_PROJECT_LOCATION: string = upath.resolve(__dirname, "./test-project");

beforeEach(() => {
    SAMPLE_CONFIGURATION = {
        entry: "./src/index.ts",
        mode: "development",
        output: {
            path: upath.resolve(TEST_PROJECT_LOCATION, "dist"),
            filename: "[name].bundle.js",
            chunkFilename: "[name].bundle.js",
            publicPath: "./"
        }
    };
});

it("Adding typescript plugin to configuration", () => {
    const configuration = new Builder(TEST_PROJECT_LOCATION, SAMPLE_CONFIGURATION).use(TypeScriptPlugin).toConfig();
    expect(configuration).toMatchSnapshot();
});

it("Adding typescript plugin with fork-ts-checker-webpack-plugin options to configuration", () => {
    const configuration = new Builder(TEST_PROJECT_LOCATION, SAMPLE_CONFIGURATION)
        .use(TypeScriptPlugin, {
            forkTsCheckerOptions: {
                formatter: "codeframe",
                silent: true,
                colors: true,
                checkSyntacticErrors: false
            }
        })
        .toConfig();

    expect(configuration).toMatchSnapshot();
});

it("Adding typescript plugin with tsconfigPathsPluginOptions to configuration that tsconfig.json do not have baseURL", () => {
    const configuration = new Builder(TEST_PROJECT_LOCATION, SAMPLE_CONFIGURATION);

    expect(() =>
        configuration
            .use(TypeScriptPlugin, {
                tsconfigPathsPluginOptions: { configFile: upath.resolve(TEST_PROJECT_LOCATION, "tsconfig.json") }
            })
            .toConfig()
    ).toThrowError("Cannot add tsconfigPathsPluginOptions because baseUrl do not exist at tsconfig.json");
});

it("Adding typescript plugin with tsconfigPathsPluginOptions to configuration", () => {
    const projectLocation = upath.resolve(__dirname, "./tsconfig-baseURL-exist");
    const configuration = new Builder(projectLocation, SAMPLE_CONFIGURATION)
        .use(TypeScriptPlugin, {
            tsconfigPathsPluginOptions: {
                logLevel: "INFO"
            }
        })
        .toConfig();

    expect(configuration).toMatchSnapshot();
});

it("baseURL exists at tsconfig", () => {
    const projectLocation = upath.resolve(__dirname, "./tsconfig-baseURL-exist");
    const configuration = new Builder(projectLocation, SAMPLE_CONFIGURATION).use(TypeScriptPlugin).toConfig();
    expect(configuration).toMatchSnapshot();
});

it("baseURL exists at tsconfig and adding tsconfigPathsPluginOptions", () => {
    const projectLocation = upath.resolve(__dirname, "./tsconfig-baseURL-exist");
    const configuration = new Builder(projectLocation, SAMPLE_CONFIGURATION)
        .use(TypeScriptPlugin, {
            tsconfigPathsPluginOptions: { configFile: upath.resolve(projectLocation, "tsconfig.json") }
        })
        .toConfig();
    expect(configuration).toMatchSnapshot();
});

it("tsconfig with single space baseUrl", () => {
    const configuration = new Builder(upath.resolve(__dirname, "./empty-space-baseURL"), SAMPLE_CONFIGURATION)
        .use(TypeScriptPlugin)
        .toConfig();
    expect(configuration).toMatchSnapshot();
});

it("tsconfig jsx property exist", () => {
    const configuration = new Builder(upath.resolve(__dirname, "./tsconfig-jsx-exist"), SAMPLE_CONFIGURATION)
        .use(TypeScriptPlugin)
        .toConfig();
    expect(configuration).toMatchSnapshot();
});

it("tsconfig and tslint do not exist", () => {
    const projectLocation = upath.resolve(__dirname, "./tsconfig-tslint-not-exist");
    fs.emptyDir(projectLocation);
    const configuration = new Builder(projectLocation, SAMPLE_CONFIGURATION).use(TypeScriptPlugin).toConfig();
    expect(configuration).toMatchSnapshot();
});
