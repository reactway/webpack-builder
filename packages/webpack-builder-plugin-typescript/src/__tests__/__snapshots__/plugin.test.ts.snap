// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`development Adding typescript plugin to configuration 1`] = `
Object {
  "devtool": "inline-source-map",
  "entry": "./src/index.ts",
  "mode": "development",
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.tsx\\?\\$/,
        "use": Array [
          Object {
            "loader": "babel-loader",
            "options": Object {
              "babelrc": true,
              "plugins": Array [
                "syntax-dynamic-import",
              ],
            },
          },
          Object {
            "loader": "ts-loader",
            "options": Object {
              "happyPackMode": true,
              "transpileOnly": true,
            },
          },
        ],
      },
      Object {
        "include": /node_modules/,
        "test": /\\\\\\.jsx\\?\\$/,
        "use": Array [
          Object {
            "loader": "babel-loader",
            "options": Object {
              "presets": Array [
                "@babel/preset-env",
              ],
            },
          },
        ],
      },
    ],
  },
  "node": Object {
    "fs": "empty",
    "net": "empty",
    "tls": "empty",
  },
  "output": Object {
    "chunkFilename": "[name].bundle.js",
    "filename": "[name].bundle.js",
    "path": "./src/__tests__///./test-project//dist",
    "publicPath": "./",
  },
  "plugins": Array [
    ForkTsCheckerWebpackPlugin {
      "apply": [MockFunction],
      "createDoneCallback": [MockFunction],
      "createEmitCallback": [MockFunction],
      "createNoopEmitCallback": [MockFunction],
      "handleServiceExit": [MockFunction],
      "handleServiceMessage": [MockFunction],
      "killService": [MockFunction],
      "pluginCompile": [MockFunction],
      "pluginDone": [MockFunction],
      "pluginEmit": [MockFunction],
      "pluginStart": [MockFunction],
      "pluginStop": [MockFunction],
      "printLoggerMessage": [MockFunction],
      "spawnService": [MockFunction],
      "validateVersions": [MockFunction],
    },
  ],
  "resolve": Object {
    "extensions": Array [
      ".ts",
      ".tsx",
      ".jsx",
      ".js",
    ],
    "plugins": Array [],
  },
}
`;

exports[`development Adding typescript plugin with fork-ts-checker-webpack-plugin options to configuration 1`] = `
Object {
  "devtool": "inline-source-map",
  "entry": "./src/index.ts",
  "mode": "development",
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.tsx\\?\\$/,
        "use": Array [
          Object {
            "loader": "babel-loader",
            "options": Object {
              "babelrc": true,
              "plugins": Array [
                "syntax-dynamic-import",
              ],
            },
          },
          Object {
            "loader": "ts-loader",
            "options": Object {
              "happyPackMode": true,
              "transpileOnly": true,
            },
          },
        ],
      },
      Object {
        "include": /node_modules/,
        "test": /\\\\\\.jsx\\?\\$/,
        "use": Array [
          Object {
            "loader": "babel-loader",
            "options": Object {
              "presets": Array [
                "@babel/preset-env",
              ],
            },
          },
        ],
      },
    ],
  },
  "node": Object {
    "fs": "empty",
    "net": "empty",
    "tls": "empty",
  },
  "output": Object {
    "chunkFilename": "[name].bundle.js",
    "filename": "[name].bundle.js",
    "path": "./src/__tests__///./test-project//dist",
    "publicPath": "./",
  },
  "plugins": Array [
    ForkTsCheckerWebpackPlugin {
      "apply": [MockFunction],
      "createDoneCallback": [MockFunction],
      "createEmitCallback": [MockFunction],
      "createNoopEmitCallback": [MockFunction],
      "handleServiceExit": [MockFunction],
      "handleServiceMessage": [MockFunction],
      "killService": [MockFunction],
      "pluginCompile": [MockFunction],
      "pluginDone": [MockFunction],
      "pluginEmit": [MockFunction],
      "pluginStart": [MockFunction],
      "pluginStop": [MockFunction],
      "printLoggerMessage": [MockFunction],
      "spawnService": [MockFunction],
      "validateVersions": [MockFunction],
    },
  ],
  "resolve": Object {
    "extensions": Array [
      ".ts",
      ".tsx",
      ".jsx",
      ".js",
    ],
    "plugins": Array [],
  },
}
`;

exports[`development Adding typescript plugin with tsconfigPathsPluginOptions to configuration 1`] = `
Object {
  "devtool": "inline-source-map",
  "entry": "./src/index.ts",
  "mode": "development",
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.tsx\\?\\$/,
        "use": Array [
          Object {
            "loader": "babel-loader",
            "options": Object {
              "babelrc": true,
              "plugins": Array [
                "syntax-dynamic-import",
              ],
            },
          },
          Object {
            "loader": "ts-loader",
            "options": Object {
              "happyPackMode": true,
              "transpileOnly": true,
            },
          },
        ],
      },
      Object {
        "include": /node_modules/,
        "test": /\\\\\\.jsx\\?\\$/,
        "use": Array [
          Object {
            "loader": "babel-loader",
            "options": Object {
              "presets": Array [
                "@babel/preset-env",
              ],
            },
          },
        ],
      },
    ],
  },
  "node": Object {
    "fs": "empty",
    "net": "empty",
    "tls": "empty",
  },
  "output": Object {
    "chunkFilename": "[name].bundle.js",
    "filename": "[name].bundle.js",
    "path": "./src/__tests__///./test-project//dist",
    "publicPath": "./",
  },
  "plugins": Array [
    ForkTsCheckerWebpackPlugin {
      "apply": [MockFunction],
      "createDoneCallback": [MockFunction],
      "createEmitCallback": [MockFunction],
      "createNoopEmitCallback": [MockFunction],
      "handleServiceExit": [MockFunction],
      "handleServiceMessage": [MockFunction],
      "killService": [MockFunction],
      "pluginCompile": [MockFunction],
      "pluginDone": [MockFunction],
      "pluginEmit": [MockFunction],
      "pluginStart": [MockFunction],
      "pluginStop": [MockFunction],
      "printLoggerMessage": [MockFunction],
      "spawnService": [MockFunction],
      "validateVersions": [MockFunction],
    },
  ],
  "resolve": Object {
    "extensions": Array [
      ".ts",
      ".js",
    ],
    "plugins": Array [
      TsconfigPathsPlugin {
        "apply": [MockFunction],
      },
    ],
  },
}
`;

exports[`development baseURL exists at tsconfig 1`] = `
Object {
  "devtool": "inline-source-map",
  "entry": "./src/index.ts",
  "mode": "development",
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.tsx\\?\\$/,
        "use": Array [
          Object {
            "loader": "babel-loader",
            "options": Object {
              "babelrc": true,
              "plugins": Array [
                "syntax-dynamic-import",
              ],
            },
          },
          Object {
            "loader": "ts-loader",
            "options": Object {
              "happyPackMode": true,
              "transpileOnly": true,
            },
          },
        ],
      },
      Object {
        "include": /node_modules/,
        "test": /\\\\\\.jsx\\?\\$/,
        "use": Array [
          Object {
            "loader": "babel-loader",
            "options": Object {
              "presets": Array [
                "@babel/preset-env",
              ],
            },
          },
        ],
      },
    ],
  },
  "node": Object {
    "fs": "empty",
    "net": "empty",
    "tls": "empty",
  },
  "output": Object {
    "chunkFilename": "[name].bundle.js",
    "filename": "[name].bundle.js",
    "path": "./src/__tests__///./test-project//dist",
    "publicPath": "./",
  },
  "plugins": Array [
    ForkTsCheckerWebpackPlugin {
      "apply": [MockFunction],
      "createDoneCallback": [MockFunction],
      "createEmitCallback": [MockFunction],
      "createNoopEmitCallback": [MockFunction],
      "handleServiceExit": [MockFunction],
      "handleServiceMessage": [MockFunction],
      "killService": [MockFunction],
      "pluginCompile": [MockFunction],
      "pluginDone": [MockFunction],
      "pluginEmit": [MockFunction],
      "pluginStart": [MockFunction],
      "pluginStop": [MockFunction],
      "printLoggerMessage": [MockFunction],
      "spawnService": [MockFunction],
      "validateVersions": [MockFunction],
    },
  ],
  "resolve": Object {
    "extensions": Array [
      ".ts",
      ".js",
    ],
    "plugins": Array [
      TsconfigPathsPlugin {
        "apply": [MockFunction],
      },
    ],
  },
}
`;

exports[`development baseURL exists at tsconfig and adding tsconfigPathsPluginOptions 1`] = `
Object {
  "devtool": "inline-source-map",
  "entry": "./src/index.ts",
  "mode": "development",
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.tsx\\?\\$/,
        "use": Array [
          Object {
            "loader": "babel-loader",
            "options": Object {
              "babelrc": true,
              "plugins": Array [
                "syntax-dynamic-import",
              ],
            },
          },
          Object {
            "loader": "ts-loader",
            "options": Object {
              "happyPackMode": true,
              "transpileOnly": true,
            },
          },
        ],
      },
      Object {
        "include": /node_modules/,
        "test": /\\\\\\.jsx\\?\\$/,
        "use": Array [
          Object {
            "loader": "babel-loader",
            "options": Object {
              "presets": Array [
                "@babel/preset-env",
              ],
            },
          },
        ],
      },
    ],
  },
  "node": Object {
    "fs": "empty",
    "net": "empty",
    "tls": "empty",
  },
  "output": Object {
    "chunkFilename": "[name].bundle.js",
    "filename": "[name].bundle.js",
    "path": "./src/__tests__///./test-project//dist",
    "publicPath": "./",
  },
  "plugins": Array [
    ForkTsCheckerWebpackPlugin {
      "apply": [MockFunction],
      "createDoneCallback": [MockFunction],
      "createEmitCallback": [MockFunction],
      "createNoopEmitCallback": [MockFunction],
      "handleServiceExit": [MockFunction],
      "handleServiceMessage": [MockFunction],
      "killService": [MockFunction],
      "pluginCompile": [MockFunction],
      "pluginDone": [MockFunction],
      "pluginEmit": [MockFunction],
      "pluginStart": [MockFunction],
      "pluginStop": [MockFunction],
      "printLoggerMessage": [MockFunction],
      "spawnService": [MockFunction],
      "validateVersions": [MockFunction],
    },
  ],
  "resolve": Object {
    "extensions": Array [
      ".ts",
      ".js",
    ],
    "plugins": Array [
      TsconfigPathsPlugin {
        "apply": [MockFunction],
      },
    ],
  },
}
`;

exports[`development tsconfig and tslint do not exist 1`] = `
Object {
  "devtool": "inline-source-map",
  "entry": "./src/index.ts",
  "mode": "development",
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.tsx\\?\\$/,
        "use": Array [
          Object {
            "loader": "babel-loader",
            "options": Object {
              "babelrc": true,
              "plugins": Array [
                "syntax-dynamic-import",
              ],
            },
          },
          Object {
            "loader": "ts-loader",
            "options": Object {
              "happyPackMode": true,
              "transpileOnly": true,
            },
          },
        ],
      },
      Object {
        "include": /node_modules/,
        "test": /\\\\\\.jsx\\?\\$/,
        "use": Array [
          Object {
            "loader": "babel-loader",
            "options": Object {
              "presets": Array [
                "@babel/preset-env",
              ],
            },
          },
        ],
      },
    ],
  },
  "node": Object {
    "fs": "empty",
    "net": "empty",
    "tls": "empty",
  },
  "output": Object {
    "chunkFilename": "[name].bundle.js",
    "filename": "[name].bundle.js",
    "path": "./src/__tests__///./test-project//dist",
    "publicPath": "./",
  },
  "plugins": Array [
    ForkTsCheckerWebpackPlugin {
      "apply": [MockFunction],
      "createDoneCallback": [MockFunction],
      "createEmitCallback": [MockFunction],
      "createNoopEmitCallback": [MockFunction],
      "handleServiceExit": [MockFunction],
      "handleServiceMessage": [MockFunction],
      "killService": [MockFunction],
      "pluginCompile": [MockFunction],
      "pluginDone": [MockFunction],
      "pluginEmit": [MockFunction],
      "pluginStart": [MockFunction],
      "pluginStop": [MockFunction],
      "printLoggerMessage": [MockFunction],
      "spawnService": [MockFunction],
      "validateVersions": [MockFunction],
    },
  ],
  "resolve": Object {
    "extensions": Array [
      ".ts",
      ".js",
    ],
    "plugins": Array [],
  },
}
`;

exports[`development tsconfig jsx property exist 1`] = `
Object {
  "devtool": "inline-source-map",
  "entry": "./src/index.ts",
  "mode": "development",
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.tsx\\?\\$/,
        "use": Array [
          Object {
            "loader": "babel-loader",
            "options": Object {
              "babelrc": true,
              "plugins": Array [
                "syntax-dynamic-import",
              ],
            },
          },
          Object {
            "loader": "ts-loader",
            "options": Object {
              "happyPackMode": true,
              "transpileOnly": true,
            },
          },
        ],
      },
      Object {
        "include": /node_modules/,
        "test": /\\\\\\.jsx\\?\\$/,
        "use": Array [
          Object {
            "loader": "babel-loader",
            "options": Object {
              "presets": Array [
                "@babel/preset-env",
              ],
            },
          },
        ],
      },
    ],
  },
  "node": Object {
    "fs": "empty",
    "net": "empty",
    "tls": "empty",
  },
  "output": Object {
    "chunkFilename": "[name].bundle.js",
    "filename": "[name].bundle.js",
    "path": "./src/__tests__///./test-project//dist",
    "publicPath": "./",
  },
  "plugins": Array [
    ForkTsCheckerWebpackPlugin {
      "apply": [MockFunction],
      "createDoneCallback": [MockFunction],
      "createEmitCallback": [MockFunction],
      "createNoopEmitCallback": [MockFunction],
      "handleServiceExit": [MockFunction],
      "handleServiceMessage": [MockFunction],
      "killService": [MockFunction],
      "pluginCompile": [MockFunction],
      "pluginDone": [MockFunction],
      "pluginEmit": [MockFunction],
      "pluginStart": [MockFunction],
      "pluginStop": [MockFunction],
      "printLoggerMessage": [MockFunction],
      "spawnService": [MockFunction],
      "validateVersions": [MockFunction],
    },
  ],
  "resolve": Object {
    "extensions": Array [
      ".ts",
      ".tsx",
      ".jsx",
      ".js",
    ],
    "plugins": Array [
      TsconfigPathsPlugin {
        "apply": [MockFunction],
      },
    ],
  },
}
`;

exports[`development tsconfig with single space baseUrl 1`] = `
Object {
  "devtool": "inline-source-map",
  "entry": "./src/index.ts",
  "mode": "development",
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.tsx\\?\\$/,
        "use": Array [
          Object {
            "loader": "babel-loader",
            "options": Object {
              "babelrc": true,
              "plugins": Array [
                "syntax-dynamic-import",
              ],
            },
          },
          Object {
            "loader": "ts-loader",
            "options": Object {
              "happyPackMode": true,
              "transpileOnly": true,
            },
          },
        ],
      },
      Object {
        "include": /node_modules/,
        "test": /\\\\\\.jsx\\?\\$/,
        "use": Array [
          Object {
            "loader": "babel-loader",
            "options": Object {
              "presets": Array [
                "@babel/preset-env",
              ],
            },
          },
        ],
      },
    ],
  },
  "node": Object {
    "fs": "empty",
    "net": "empty",
    "tls": "empty",
  },
  "output": Object {
    "chunkFilename": "[name].bundle.js",
    "filename": "[name].bundle.js",
    "path": "./src/__tests__///./test-project//dist",
    "publicPath": "./",
  },
  "plugins": Array [
    ForkTsCheckerWebpackPlugin {
      "apply": [MockFunction],
      "createDoneCallback": [MockFunction],
      "createEmitCallback": [MockFunction],
      "createNoopEmitCallback": [MockFunction],
      "handleServiceExit": [MockFunction],
      "handleServiceMessage": [MockFunction],
      "killService": [MockFunction],
      "pluginCompile": [MockFunction],
      "pluginDone": [MockFunction],
      "pluginEmit": [MockFunction],
      "pluginStart": [MockFunction],
      "pluginStop": [MockFunction],
      "printLoggerMessage": [MockFunction],
      "spawnService": [MockFunction],
      "validateVersions": [MockFunction],
    },
  ],
  "resolve": Object {
    "extensions": Array [
      ".ts",
      ".js",
    ],
    "plugins": Array [],
  },
}
`;

exports[`production Adding typescript plugin to configuration in production mode 1`] = `
Object {
  "devtool": "inline-source-map",
  "entry": "./src/index.ts",
  "mode": "production",
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.tsx\\?\\$/,
        "use": Array [
          Object {
            "loader": "babel-loader",
            "options": Object {
              "babelrc": true,
              "plugins": Array [
                "syntax-dynamic-import",
              ],
            },
          },
          Object {
            "loader": "ts-loader",
            "options": Object {
              "happyPackMode": true,
              "transpileOnly": true,
            },
          },
        ],
      },
      Object {
        "include": /node_modules/,
        "test": /\\\\\\.jsx\\?\\$/,
        "use": Array [
          Object {
            "loader": "babel-loader",
            "options": Object {
              "presets": Array [
                "@babel/preset-env",
              ],
            },
          },
        ],
      },
    ],
  },
  "node": Object {
    "fs": "empty",
    "net": "empty",
    "tls": "empty",
  },
  "optimization": Object {
    "minimizer": Array [
      TerserPlugin {
        "apply": [MockFunction],
      },
    ],
  },
  "output": Object {
    "chunkFilename": "[name].bundle.js",
    "filename": "[name].bundle.js",
    "path": "./src/__tests__///./test-project//dist",
    "publicPath": "./",
  },
  "plugins": Array [
    ForkTsCheckerWebpackPlugin {
      "apply": [MockFunction],
      "createDoneCallback": [MockFunction],
      "createEmitCallback": [MockFunction],
      "createNoopEmitCallback": [MockFunction],
      "handleServiceExit": [MockFunction],
      "handleServiceMessage": [MockFunction],
      "killService": [MockFunction],
      "pluginCompile": [MockFunction],
      "pluginDone": [MockFunction],
      "pluginEmit": [MockFunction],
      "pluginStart": [MockFunction],
      "pluginStop": [MockFunction],
      "printLoggerMessage": [MockFunction],
      "spawnService": [MockFunction],
      "validateVersions": [MockFunction],
    },
  ],
  "resolve": Object {
    "extensions": Array [
      ".ts",
      ".tsx",
      ".jsx",
      ".js",
    ],
    "plugins": Array [],
  },
}
`;

exports[`production Adding typescript plugin to configuration in production mode and passsing terser options 1`] = `
Object {
  "devtool": "inline-source-map",
  "entry": "./src/index.ts",
  "mode": "production",
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.tsx\\?\\$/,
        "use": Array [
          Object {
            "loader": "babel-loader",
            "options": Object {
              "babelrc": true,
              "plugins": Array [
                "syntax-dynamic-import",
              ],
            },
          },
          Object {
            "loader": "ts-loader",
            "options": Object {
              "happyPackMode": true,
              "transpileOnly": true,
            },
          },
        ],
      },
      Object {
        "include": /node_modules/,
        "test": /\\\\\\.jsx\\?\\$/,
        "use": Array [
          Object {
            "loader": "babel-loader",
            "options": Object {
              "presets": Array [
                "@babel/preset-env",
              ],
            },
          },
        ],
      },
    ],
  },
  "node": Object {
    "fs": "empty",
    "net": "empty",
    "tls": "empty",
  },
  "optimization": Object {
    "minimizer": Array [
      TerserPlugin {
        "apply": [MockFunction],
      },
    ],
  },
  "output": Object {
    "chunkFilename": "[name].bundle.js",
    "filename": "[name].bundle.js",
    "path": "./src/__tests__///./test-project//dist",
    "publicPath": "./",
  },
  "plugins": Array [
    ForkTsCheckerWebpackPlugin {
      "apply": [MockFunction],
      "createDoneCallback": [MockFunction],
      "createEmitCallback": [MockFunction],
      "createNoopEmitCallback": [MockFunction],
      "handleServiceExit": [MockFunction],
      "handleServiceMessage": [MockFunction],
      "killService": [MockFunction],
      "pluginCompile": [MockFunction],
      "pluginDone": [MockFunction],
      "pluginEmit": [MockFunction],
      "pluginStart": [MockFunction],
      "pluginStop": [MockFunction],
      "printLoggerMessage": [MockFunction],
      "spawnService": [MockFunction],
      "validateVersions": [MockFunction],
    },
  ],
  "resolve": Object {
    "extensions": Array [
      ".ts",
      ".tsx",
      ".jsx",
      ".js",
    ],
    "plugins": Array [],
  },
}
`;
